### ISSUE TITLE
WVD Pool Rebuild Issue

### ISSUE THREAD

**Issue by Digiroka** [Azure/RDS-Templates#259]
Hi, I am attempting to rebuild a WVD pool - I am using a service principal for this task, which has been successfully used to create pools in the past.
The rebuild fails with the below issue. 

`The received access token is not valid: at least one of the claims 'puid' or 'altsecid' or 'oid' should be present. If you are accessing as application please make sure service principal is properly created in the tenant`

Full error below, any help appreciated. Thanks.

`{"code":"DeploymentFailed","message":"At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/arm-debug for usage details.","details":[{"code":"Conflict","message":"{\r\n \"status\": \"Failed\",\r\n \"error\": {\r\n \"code\": \"ResourceDeploymentFailure\",\r\n \"message\": \"The resource operation completed with terminal provisioning state 'Failed'.\",\r\n \"details\": [\r\n {\r\n \"code\": \"VMExtensionProvisioningError\",\r\n \"message\": \"VM has reported a failure when processing extension 'dscextension'. Error message: \\\"DSC Configuration 'FirstSessionHost' completed with error(s). Following are the first few: PowerShell DSC resource MSFT_ScriptResource failed to execute Set-TargetResource functionality with error message: The received access token is not valid: at least one of the claims 'puid' or 'altsecid' or 'oid' should be present. If you are accessing as application please make sure service principal is properly created in the tenant. The SendConfigurationApply function did not succeed.\\\".\"\r\n }\r\n ]\r\n }\r\n}"}]}`

**Comment by Digiroka** [comment]
Update - I've just run the same process using an AAD account with RDS Owner / Tenant Creator permissions, and the task has completed without issue.

Can anyone suggest why it didn't work with the Service Principal? This account has been used to create multiple host pools, including the one being updated.

Thanks

**Comment by ChristianMontoya** [comment]
@Digiroka : Which steps did you use to create the service principal?

**Comment by Digiroka** [comment]
@ChristianMontoya 
I used the below process, and have successfully used this SP to create host pools via the gui previously...
https://docs.microsoft.com/en-us/azure/virtual-desktop/create-service-principal-role-powershell

**Comment by Digiroka** [comment]
@ChristianMontoya please can this be classified as "2-Update-existing-WVD-host-pool"

**Comment by ChristianMontoya** [comment]
@Digiroka : Can you run an additional test for us? Can you create a new service principal but make it single-tenant instead of multi-tenant:
`$svcPrincipal = New-AzureADApplication -AvailableToOtherTenants $false -DisplayName "Windows Virtual Desktop Svc Principal"`

Then, re-test the Create or Update? We're adding support for single-tenant, which should make these templates more reliable.

**Comment by EirikVI** [comment]
> @Digiroka : Can you run an additional test for us? Can you create a new service principal but make it single-tenant instead of multi-tenant:
> `$svcPrincipal = New-AzureADApplication -AvailableToOtherTenants $false -DisplayName "Windows Virtual Desktop Svc Principal"`
> 
> Then, re-test the Create or Update? We're adding support for single-tenant, which should make these templates more reliable.

@ChristianMontoya: I did what you suggested. Same error.

**Comment by Digiroka** [comment]
@ChristianMontoya I also have the same error.

**Comment by ChristianMontoya** [comment]
@lintFan : Let's try and take a look at this this week.

**Comment by lintFan** [comment]
@Digiroka  @EirikVI 
There was an update #317  that was merged to the master branch on Feb 3, 2020 12:54 PM PST that updated the [Microsoft.RDInfra.RDPowershell powershell module](https://docs.microsoft.com/en-us/powershell/windows-virtual-desktop/overview) that it uses from version 1.0.1.11 to 1.0.1534.2001, this includes fixes for some bugs that existed around authenticating using service principles. So I think we should first check that you are using a version that has this change.  Note that this change was only deployed to the Azure Marketplace more recently at about Feb 26, 2020 3:40 AM PT.

Are you running this using the Azure Marketplace https://ms.portal.azure.com/#create/rds.wvd-provision-host-poolpreview or using the code from this github repo using  New-AzResourceGroupDeployment? The original error message posted in this issue in on Nov 14, 2019 is from older code because it contains
`DSC Configuration 'FirstSessionHost' completed with error(s)`
and FirstSessionHost is no longer used in the current code. Are you still getting the exact same error message as the one posted on Nov 14, 2019? If you are then when did you run it last? If you are not then please post the new error message.

If you are choosing the option to redeploy a deployment from the Azure Portal where the mainTemplate.json references this github.com repository then it may be using the same version of this github code that was used in the initial deployment, to use the newer code you will need to do a new deployment.

If you are running a local copy of the github code you check that it has this change by extracting \wvd-templates\DSC\Configuration.zip and then inside that extracting PowerShellModules.zip and then check that the properties of the file Microsoft.RDInfra.RDPowershell.dll has a File version of 1.0.1535.0 (Note that the version number of the Microsoft.RDInfra.RDPowershell listed in the [powershell gallery](https://www.powershellgallery.com/packages/Microsoft.RDInfra.RDPowershell)  is different than the file version number listed in the properties of Microsoft.RDInfra.RDPowershell.dll.)

**Comment by Digiroka** [comment]
Hello, I built a new test pool this morning using the gallery image, then ran the Update Hostpool deplyment directly from the git link: https://github.com/Azure/RDS-Templates/tree/master/wvd-templates/Update%20existing%20WVD%20host%20pool

When I ran the deployment with 'action on previous machines' set to 'delete', the task completed without error - however it did not delete the original VM, it just updated the VM extension.

When I ran the deployment with the 'Deprovision' action for 'action on previous machines', I had the same result. the task completed without issue, bit the original machine had not been deprovisioned.

So basically, no chages were made to the session host in either scenario, and no errors were shown in the deployment.

**Comment by lintFan** [comment]
@nakranimohit0 thoughts on this?

**Comment by nakranimohit0** [comment]
Are you sure you are using a different `rdshNamePrefix` for update template than the one used for create template ? If so, do you see the new VMs with the new name or same as before ? Can you check the script log as mentioned here: https://docs.microsoft.com/en-us/azure/virtual-desktop/troubleshoot-vm-configuration#windows-virtual-desktop-agent-and-windows-virtual-desktop-boot-loader-are-not-installed ?

**Comment by Digiroka** [comment]
@nakranimohit0 - thank you, I was using the same rdsnameprefix, I will try with a new prefix as requested and report back.  
This requirement however seems counterintuitive. I would normally expect to rebuild a pool with the same naming convention. Could this ability possibly be introduced?

**Comment by Digiroka** [comment]
@nakranimohit0 Using a new 'rdsnameprefix' per-deployment, below are my results.

**Using AAD account with RDS Owner:**

Rebuild and **Deprovision** old Session Hosts - Success. Old hosts removed from pool and deprovisioned.
Rebuild and **Delete** old Session Hosts - Success. Old hosts removed from pool and deleted - Availability set not deleted.

Using Service Principal (RDS Owner):

build and **Deprovision** old Session Hosts - Failure. Error below.

'{"code":"DeploymentFailed","message":"At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/DeployOperations for usage details.","details":[{"code":"Conflict","message":"{\r\n \"status\": \"Failed\",\r\n \"error\": {\r\n \"code\": \"ResourceDeploymentFailure\",\r\n \"message\": \"The resource operation completed with terminal provisioning state 'Failed'.\",\r\n \"details\": [\r\n {\r\n \"code\": \"VMExtensionProvisioningError\",\r\n \"message\": \"VM has reported a failure when processing extension 'dscextension'. Error message: \\\"DSC Configuration 'RegisterSessionHostAndCleanup' completed with error(s). Following are the first few: PowerShell DSC resource MSFT_ScriptResource failed to execute Set-TargetResource functionality with error message: The received access token is not valid: at least one of the claims 'puid' or 'altsecid' or 'oid' should be present. If you are accessing as application please make sure service principal is properly created in the tenant. The SendConfigurationApply function did not succeed.\\\"\\r\\n\\r\\nMore information on troubleshooting is available at https://aka.ms/VMExtensionDSCWindowsTroubleshoot \"\r\n }\r\n ]\r\n }\r\n}"}]}'

Rebuild and **Delete** old Session Hosts - Failure. Error below.

'{"code":"DeploymentFailed","message":"At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/DeployOperations for usage details.","details":[{"code":"Conflict","message":"{\r\n \"status\": \"Failed\",\r\n \"error\": {\r\n \"code\": \"ResourceDeploymentFailure\",\r\n \"message\": \"The resource operation completed with terminal provisioning state 'Failed'.\",\r\n \"details\": [\r\n {\r\n \"code\": \"VMExtensionProvisioningError\",\r\n \"message\": \"VM has reported a failure when processing extension 'dscextension'. Error message: \\\"DSC Configuration 'RegisterSessionHostAndCleanup' completed with error(s). Following are the first few: PowerShell DSC resource MSFT_ScriptResource failed to execute Set-TargetResource functionality with error message: The received access token is not valid: at least one of the claims 'puid' or 'altsecid' or 'oid' should be present. If you are accessing as application please make sure service principal is properly created in the tenant. The SendConfigurationApply function did not succeed.\\\"\\r\\n\\r\\nMore information on troubleshooting is available at https://aka.ms/VMExtensionDSCWindowsTroubleshoot \"\r\n }\r\n ]\r\n }\r\n}"}]}'

**Comment by Deofex** [comment]
For the record: I'm getting the same results as Digiroka in multiple environments. (don't want to hijack the issue, just wanted to let you know, it's not only his re-deployment)

**Comment by nakranimohit0** [comment]
@Digiroka can you please share the script log located in the VM @ C:\Windows\Temp\ScriptLog.log ?
You can mask/clear any sensitive info from the log file, but please share the log for the first VM-0 created by the [create template](https://github.com/Azure/RDS-Templates/tree/master/wvd-templates/Create%20and%20provision%20WVD%20host%20pool) and the last NEW-VM-n created by the [update template](https://github.com/Azure/RDS-Templates/tree/master/wvd-templates/Create%20and%20provision%20WVD%20host%20pool) ?

Also please share the parameters used for both templates create & update.

**Comment by nakranimohit0** [comment]
Closing out this issue as a duplicate to #321 
Please re-open if its diff.

**Comment by robinhobo** [comment]
Thanks @nakranimohit0, @ChristianMontoya and the rest of the team for your support! I have created a step-by-step blog for the solution : https://www.robinhobo.com/how-to-create-an-azure-service-principal-for-use-with-windows-virtual-desktop-and-azure-arm-templates-like-the-arm-template-to-update-an-existing-windows-virtual-desktop-hostpool/
